File: data-reader.py
#! /usr/bin/python

import re
from collections import defaultdict

def read_file_lines(input_file):
    with open(input_file, "r") as file:
        input = file.readlines()
    return input

def get_body_lines(input, start_pat, end_pat):
    mstart = re.compile(start_pat)
    mend = re.compile(end_pat)
    start = -1
    end = -1
    for i in range(len(input)):
        if mstart.match(input[i]):
            start = i
        elif mend.match(input[i]):
            end = i
            break
    return input[start : end]

def get_purchases(data):
    purchase_pattern = re.compile("^\s*([\.0-9]+)\s+([\.0-9]+)\s+([_0-9]+)\s+(\S+)\s+(\"[^\"]+\")(\s+#.*)?$")
    sec_map = defaultdict(list) # str -> list[tuple[float]]
    for line in data:
        match = purchase_pattern.match(line)
        if match:
            price = float(match.group(1))
            pieces = float(match.group(2))
            date = match.group(3)
            account = match.group(4)
            security = match.group(5)
            sec_map[security].append((price, pieces))
    return sec_map

############################################################
# TODO: Remaining code/behavior has not been documented
############################################################
# list[string] -> ( string -> tuple[string, float] )
def get_references(input):
    date_match = re.match("^\s*([_0-9]+)$", input[0])
    if date_match:
        date = date_match.group(1)
    else:
        date = "" # TODO: handle properly
    references = dict()
    for i in range(1, len(input)):
        ref_res = format_reference_line(input[i])
        references[ref_res[0]] = (date, ref_res[1])
    return references

def format_reference_line(line):
    match = re.match("^\s*(\"[^\"]+\")\s+([\.0-9]+)$", line)
    if match:
        return (match.group(1), float(match.group(2)))
    else:
        return ()

# sec_map    : "a" -> [(1.00, 10), (1.01, 10), (1.02, 10), (1.03, 10)]
# references : "a" -> (190331_0000, 1.04)

def security_principal(purchases):
    sec_principal = 0
    for purchase in purchases:
        sec_principal += purchase[0] * purchase[1]
    return sec_principal

def portfolio_principal(portfolio):
    pf_principal = 0
    for security in portfolio:
        pf_principal += security_principal(portfolio[security])
    return pf_principal

def portfolio_value(portfolio, prices):
    pf_val = 0
    for security in portfolio:
        pf_val += security_value(portfolio[security], prices[security][1])
    return pf_val

def security_value(purchases, price):
    sec_val = 0
    for purchase in purchases:
        sec_val += purchase[1]
    return sec_val * price

def make_security_report(sec_name, sec_map, references):
    sec_price = references[sec_name][1]
    sec_principal = security_principal(sec_map[sec_name])
    sec_val = security_value(sec_map[sec_name], sec_price)
    diff = sec_val - sec_principal
    percent_gain = 100 * diff / sec_principal
    return (sec_val, percent_gain, "%")

def make_portfolio_report(sec_map, references, fmt):
    pf_val = portfolio_value(sec_map, references)
    pf_principal = portfolio_principal(sec_map)
    pf_diff = pf_val - pf_principal
    pf_percent_gain = 100 * pf_diff / pf_principal
    return fmt % ("Portfolio", pf_val, pf_percent_gain, "%")

def make_full_report(sec_map, references, fmt):
    report = []
    diffs = []
    for key in sec_map:
        report.append(fmt % ((key,) + make_security_report(key, sec_map, references)))
        diff = security_value(sec_map[key], references[key][1]) -\
                security_principal(sec_map[key])
        diffs.append(diff)

    abs_diffs = map(lambda x: abs(x), diffs)
    abs_pf_diff = reduce(lambda x, y: x + y, abs_diffs)
    for i in range(len(abs_diffs)):
        frac = 100 * abs_diffs[i] / abs_pf_diff
        report[i] += " (%5.2f%s of total swing)" % (frac, "%")

    report.append('-' * 35)
    report.append(make_portfolio_report(sec_map, references, fmt))
    return report

def main():
    input = read_file_lines("dummy-data.txt")
    data = get_body_lines(input, "^tab_start:$", "^tab_end$")
    data = data[2 : ]
    sec_map = get_purchases(data)

    ref_data = get_body_lines(input, "^ref_start:$", "^ref_end$")
    references = get_references(ref_data[1 : ])

    fmt = "%15s %9.2f, %7.2f%s"
    final_report = '\n'.join(make_full_report(sec_map, references, fmt))
    print final_report
    print "OK" if test(final_report) else "error"

def test(report):
    key = """            "a"     41.60,    2.46% (20.00% of total swing)
            "b"     92.40,    1.65% (30.00% of total swing)
            "c"    204.00,    1.24% (50.00% of total swing)
-----------------------------------
      Portfolio    338.00,    1.50%"""
    return report == key

main()

File: DataReader.scala
import scala.collection.mutable.Map

object DataReader {
  def read_file_lines(filename: String): Array[String] = {
    val src = io.Source.fromFile(filename)
    try src.mkString.split("\n") finally src.close()
  }

  def get_body_lines(input: Array[String], start_pat: String, end_pat: String) = {
    val mstart = start_pat.r
    val mend = end_pat.r
    var start = -1 // use Option[Int] = Option.empty instead ?
    var end = -1   // use Option[Int] = Option.empty instead ?
    input.zipWithIndex.foreach {
      case (mstart(_*), i) => start = i
      case (mend(_*), i) => end = i
      case _ =>
    }
    input.slice(start, end)
  }

  def get_purchases(data: Array[String]) = {
    val purchase_pattern =   """\s*([\.0-9]+)\s+([\.0-9]+)\s+([_0-9]+)\s+(\S+)\s+("[^"]+")(\s+#.*)?""".r
    val sec_map = Map[String, List[(Double, Double)]]()
    for (line <- data) {
      line match {
        case purchase_pattern(price, pieces, _, _, security, _) => {
          val currList = sec_map.getOrElse(security, List[(Double, Double)]())
          sec_map += (security -> ((price.toDouble, pieces.toDouble) :: currList))
        }
        case _ => println(s"not match for line $line")
      }
    }
    sec_map
  }

  def get_references(input: Array[String]) = {
    val references = Map[String, List[(String, Double)]]()
    val dates = input(0).split("\\s+").filter(d => d.length != 0)
    val refData = input.tail
    for (i <- 0 until refData.length) {
      val line = refData(i)
      val lastQuote = line.lastIndexOf('"')
      val firstField = line.substring(0, lastQuote + 1)
      val fields = line.substring(lastQuote + 1).split("\\s+").filter(d => d.length != 0)
      val tups = for (i <- 0 until dates.length) yield {
        (dates(i), fields(i).toDouble)
      }
      references += (firstField -> tups.toList)
    }
    references
  }

  def getLatestReferences(references: Map[String, List[(String, Double)]]) = {
    val latest = Map[String, (String, Double)]()
    for ((k,l) <- references) {
      val sortedList = l.sortWith(_._1 > _._1)
      latest += (k -> sortedList(0))
    }
    latest
  }

  def format_reference_line(line: String) = {
    val match0 = """\s*("[^"]+")\s+([\.0-9]+)(\s+.*)?""".r
    line match {
      case match0(name, price, _) => (name, price.toDouble)
      case _ => ("", 0.0)
    }
  }

  def make_full_report(sec_map:    Map[String, List[(Double, Double)]],
                       references: Map[String, (String, Double)],
                       fmt:        String) = {
    import scala.collection.mutable.MutableList // :-(
    var diffs = MutableList[Double]()
    var fullReport = MutableList[String]()
    for (key <- sec_map.keys) {
      val currSecMap = sec_map(key)
      val currReference = references(key)
      val secReport = make_security_report(currSecMap, currReference)
      val diff = security_value(currSecMap, currReference._2) - security_principal(currSecMap)
      fullReport += fmt.format(key, secReport._1, secReport._2, "%")
      diffs += diff
    }

    val abs_diffs = diffs.map(x => Math.abs(x))
    val abs_pf_diff = abs_diffs.reduce((x,y) => x + y)
    for (i <- 0 until abs_diffs.length) {
      val frac = 100 * abs_diffs(i) / abs_pf_diff
      fullReport(i) = fullReport(i) + " (%5.2f%s of total swing)".format(frac, "%")
    }

    val pfReport = make_portfolio_report(sec_map, references)
    fullReport.sortWith(_ < _) :+ "-----------------------------------" :+
      fmt.format("Portfolio", pfReport._1, pfReport._2, "%")
  }

  def make_security_report(sec_map:    List[(Double, Double)],
                           references: (String, Double)) = {
    val sec_price = references._2
    val sec_principal = security_principal(sec_map)
    val sec_val = security_value(sec_map, sec_price)
    val diff = sec_val - sec_principal
    val percent_gain = 100 * diff / sec_principal
    (sec_val, percent_gain)
  }

  def security_principal(purchases: List[(Double, Double)]) = {
    purchases.map { case (a, b) => a * b }.sum
  }

  def security_value(purchases: List[(Double, Double)], price: Double) = {
    price * purchases.map { case (a, b) => b }.sum
  }

  def make_portfolio_report(sec_map:    Map[String, List[(Double, Double)]],
                            references: Map[String, (String, Double)]) = {
    val pf_val = portfolio_value(sec_map, references)
    val pf_principal = portfolio_principal(sec_map)
    val pf_diff = pf_val - pf_principal
    val pf_percent_gain = 100 * pf_diff / pf_principal
    (pf_val, pf_percent_gain)
  }

  def portfolio_value(portfolio: Map[String, List[(Double, Double)]],
                      prices:    Map[String, (String, Double)]) = {
    var pf_val = 0.0
    for (security <- portfolio.keys)
      pf_val += security_value(portfolio(security), prices(security)._2)
    pf_val
  }

  def portfolio_principal(portfolio: Map[String, List[(Double, Double)]]) = {
    var pf_principal = 0.0
    for (security <- portfolio.keys)
      pf_principal += security_principal(portfolio(security))
    pf_principal
  }

  def makeDateHeader(refData: Array[String]) = {
    val dates = refData(0).split("\\s+").filter(d => d.length != 0)
    " " * (15 + 2) + dates.map(s => "%11s".format(s)).mkString("  ")
  }

  def makeAnotherReport(sec_map:    Map[String, List[(Double, Double)]],
                        references: Map[String, List[(String, Double)]]) = {
    for ((k, l) <- references) yield {
      val sec_principal = security_principal(sec_map(k))
      val percs = for (price <- l) yield {
        val secVal = security_value(sec_map(k), price._2);
        100 * (secVal - sec_principal) / sec_principal
      }
      "%15s  ".format(k) + percs.map(p => "%10.2f%s".format(p, "%")).mkString("  ")
    }
  }

  def test(report: String) = {
    val key = """            "a"     42.00,    3.45% (20.00% of total swing)
            "b"     93.00,    2.31% (30.00% of total swing)
            "c"    205.00,    1.74% (50.00% of total swing)
-----------------------------------
      Portfolio    340.00,    2.10%"""
    report == key
  }

  def main(args: Array[String]) {
    val doTest = args.length < 1
    val inputFile = if (!doTest) args(0) else "dummy-data.txt"
    val input = read_file_lines(inputFile)

    var data = get_body_lines(input, "tab_start:", "tab_end");
    val sec_map = get_purchases(data.slice(2, data.length))

    val ref_data = get_body_lines(input, "ref_start:", "ref_end")
    val references = get_references(ref_data.slice(1, ref_data.length))
    val latestRefs = getLatestReferences(references)

    val fmt = "%15s %9.2f, %7.2f%s"
    val final_report = make_full_report(sec_map, latestRefs, fmt).mkString("\n")
    println(final_report)
    if (doTest) {
      if (test(final_report))
        println("OK")
      else
        println("error")
    }

    println()

    val dateHeader = makeDateHeader(ref_data.tail)
    val foobar = makeAnotherReport(sec_map, references)
    val foolist = dateHeader :: "-" * dateHeader.length :: foobar.toList.sortWith(_ < _)
    println(foolist.mkString("\n"))
  }
}
